File: go/sandbox.go
package clarian

import (
	"context"
	"fmt"
	"net/http"
	"net/url"
	"strings"
)

// Magic PIX keys honored by the sandbox payout rail.
const (
	SandboxFailPixKey    = "fail@sandbox.clarian"
	SandboxPendingPixKey = "pending@sandbox.clarian"
)

// Sample webhook event types for sandbox testing (exact gateway strings).
const (
	EventPixPayinCreated    = "pix_payin.created"
	EventPixPayinCompleted  = "pix_payin.completed"
	EventPixPayinExpired    = "pix_payin.expired"
	EventPixPayoutCreated   = "pix_payout.created"
	EventPixPayoutCompleted = "pix_payout.completed"
	EventPixPayoutFailed    = "pix_payout.failed"
	EventCheckoutPaid       = "checkout.paid"
)

// SandboxService exposes sandbox-only test helpers. Server-side these endpoints
// return 404 {error:"sandbox_only"} outside the test environment; the client
// also refuses non-cl_test_sk_ keys before any HTTP call.
type SandboxService struct{ c *Client }

func (s *SandboxService) requireTestKey() error {
	if !strings.HasPrefix(s.c.apiKey, "cl_test_sk_") {
		return fmt.Errorf("clarian: sandbox helpers require a cl_test_sk_ key")
	}
	return nil
}

// SimulateCashIn advances a sandbox PIX pay-in to status (completed, expired,
// or failed). Empty status defaults to completed.
// POST /pix/payins/{id}/simulate
func (s *SandboxService) SimulateCashIn(ctx context.Context, id, status string) (*CashInOrderResponse, error) {
	if err := s.requireTestKey(); err != nil {
		return nil, err
	}
	if status == "" {
		status = "completed"
	}
	switch status {
	case "completed", "expired", "failed":
	default:
		return nil, fmt.Errorf("clarian: invalid cash-in simulate status %q", status)
	}
	var out CashInOrderResponse
	body := map[string]string{"status": status}
	if err := s.c.do(ctx, http.MethodPost, "/pix/payins/"+url.PathEscape(id)+"/simulate", "", body, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// SimulateCashOut advances a sandbox PIX pay-out to status (completed or failed).
// Empty status defaults to completed.
// POST /pix/payouts/{id}/simulate
func (s *SandboxService) SimulateCashOut(ctx context.Context, id, status string) (*CashOutOrderResponse, error) {
	if err := s.requireTestKey(); err != nil {
		return nil, err
	}
	if status == "" {
		status = "completed"
	}
	switch status {
	case "completed", "failed":
	default:
		return nil, fmt.Errorf("clarian: invalid cash-out simulate status %q", status)
	}
	var out CashOutOrderResponse
	body := map[string]string{"status": status}
	if err := s.c.do(ctx, http.MethodPost, "/pix/payouts/"+url.PathEscape(id)+"/simulate", "", body, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// SandboxWebhookEventResult is returned after enqueueing a sandbox test delivery.
type SandboxWebhookEventResult struct {
	OK          bool
	Environment Environment
	EventType   string
	Enqueued    int
	DeliveryID  string // empty when the gateway returns null
}

// SendWebhookEvent enqueues a sample webhook delivery for a subscription.
// POST /webhooks/{id}/test
func (s *SandboxService) SendWebhookEvent(ctx context.Context, subscriptionID, eventType string) (*SandboxWebhookEventResult, error) {
	if err := s.requireTestKey(); err != nil {
		return nil, err
	}
	var res struct {
		OK          bool        `json:"ok"`
		Environment Environment `json:"environment"`
		EventType   string      `json:"event_type"`
		Enqueued    struct {
			Enqueued   int     `json:"enqueued"`
			DeliveryID *string `json:"delivery_id"`
		} `json:"enqueued"`
	}
	body := map[string]string{"event_type": eventType}
	if err := s.c.do(ctx, http.MethodPost, "/webhooks/"+url.PathEscape(subscriptionID)+"/test", "", body, &res); err != nil {
		return nil, err
	}
	out := &SandboxWebhookEventResult{
		OK:          res.OK,
		Environment: res.Environment,
		EventType:   res.EventType,
		Enqueued:    res.Enqueued.Enqueued,
	}
	if res.Enqueued.DeliveryID != nil {
		out.DeliveryID = *res.Enqueued.DeliveryID
	}
	return out, nil
}

// ResendWebhookDelivery re-enqueues an existing delivery and returns the new
// delivery id. POST /webhooks/deliveries/{id}/resend
func (s *SandboxService) ResendWebhookDelivery(ctx context.Context, deliveryID string) (string, error) {
	if err := s.requireTestKey(); err != nil {
		return "", err
	}
	var res struct {
		OK         bool   `json:"ok"`
		DeliveryID string `json:"delivery_id"`
	}
	if err := s.c.do(ctx, http.MethodPost, "/webhooks/deliveries/"+url.PathEscape(deliveryID)+"/resend", "", nil, &res); err != nil {
		return "", err
	}
	return res.DeliveryID, nil
}

File: php/src/Sandbox.php
<?php

declare(strict_types=1);

namespace Clarian;

/**
 * Sandbox-only test helpers.
 *
 * Server-side these endpoints return 404 {error:"sandbox_only"} outside the
 * test environment; the client also refuses non-cl_test_sk_ keys before any
 * HTTP call (throws LogicException).
 */
final class Sandbox
{
    public const SANDBOX_FAIL_PIX_KEY = 'fail@sandbox.clarian';
    public const SANDBOX_PENDING_PIX_KEY = 'pending@sandbox.clarian';

    public const EVENT_PIX_PAYIN_CREATED = 'pix_payin.created';
    public const EVENT_PIX_PAYIN_COMPLETED = 'pix_payin.completed';
    public const EVENT_PIX_PAYIN_EXPIRED = 'pix_payin.expired';
    public const EVENT_PIX_PAYOUT_CREATED = 'pix_payout.created';
    public const EVENT_PIX_PAYOUT_COMPLETED = 'pix_payout.completed';
    public const EVENT_PIX_PAYOUT_FAILED = 'pix_payout.failed';
    public const EVENT_CHECKOUT_PAID = 'checkout.paid';

    private const CASH_IN_STATUSES = ['completed', 'expired', 'failed'];
    private const CASH_OUT_STATUSES = ['completed', 'failed'];

    public function __construct(private readonly Client $client)
    {
    }

    private function requireTestKey(): void
    {
        if (!str_starts_with($this->client->apiKey, 'cl_test_sk_')) {
            throw new \LogicException('sandbox helpers require a cl_test_sk_ key');
        }
    }

    /**
     * POST /pix/payins/{id}/simulate.
     * Status: completed | expired | failed. Empty/null defaults to completed.
     *
     * @return array<string, mixed>
     */
    public function simulateCashIn(string $orderId, ?string $status = null): array
    {
        $this->requireTestKey();
        $resolved = ($status === null || $status === '') ? 'completed' : $status;
        if (!in_array($resolved, self::CASH_IN_STATUSES, true)) {
            throw new \InvalidArgumentException(
                'invalid cash-in simulate status "' . $resolved . '"',
            );
        }

        return $this->client->request(
            'POST',
            '/pix/payins/' . Client::pathEscape($orderId) . '/simulate',
            ['status' => $resolved],
        );
    }

    /**
     * POST /pix/payouts/{id}/simulate.
     * Status: completed | failed. Empty/null defaults to completed.
     *
     * @return array<string, mixed>
     */
    public function simulateCashOut(string $orderId, ?string $status = null): array
    {
        $this->requireTestKey();
        $resolved = ($status === null || $status === '') ? 'completed' : $status;
        if (!in_array($resolved, self::CASH_OUT_STATUSES, true)) {
            throw new \InvalidArgumentException(
                'invalid cash-out simulate status "' . $resolved . '"',
            );
        }

        return $this->client->request(
            'POST',
            '/pix/payouts/' . Client::pathEscape($orderId) . '/simulate',
            ['status' => $resolved],
        );
    }

    /**
     * POST /webhooks/{id}/test: enqueue a sample delivery.
     *
     * @return array{
     *   ok: mixed,
     *   environment: mixed,
     *   event_type: mixed,
     *   enqueued: int,
     *   delivery_id: string
     * }
     */
    public function sendWebhookEvent(string $subscriptionId, string $eventType): array
    {
        $this->requireTestKey();
        $res = $this->client->request(
            'POST',
            '/webhooks/' . Client::pathEscape($subscriptionId) . '/test',
            ['event_type' => $eventType],
        );
        $enqueued = $res['enqueued'] ?? [];
        if (!is_array($enqueued)) {
            $enqueued = [];
        }
        $deliveryId = $enqueued['delivery_id'] ?? '';

        return [
            'ok' => $res['ok'] ?? null,
            'environment' => $res['environment'] ?? null,
            'event_type' => $res['event_type'] ?? null,
            'enqueued' => (int) ($enqueued['enqueued'] ?? 0),
            'delivery_id' => $deliveryId !== null ? (string) $deliveryId : '',
        ];
    }

    /**
     * POST /webhooks/deliveries/{id}/resend; returns the new delivery id.
     */
    public function resendWebhookDelivery(string $deliveryId): string
    {
        $this->requireTestKey();
        $res = $this->client->request(
            'POST',
            '/webhooks/deliveries/' . Client::pathEscape($deliveryId) . '/resend',
        );

        return (string) ($res['delivery_id'] ?? '');
    }
}
